strax/corrections.py

```diff
@@ -160,25 +183,60 @@ def write(self, correction, df, required_columns=('ONLINE', 'v1')):
         for req in required_columns:
             if req not in df.columns:
                 raise ValueError(f'Must specify {req} in dataframe')
-        # Compare against
+        # Compare against existing data
+        # We can change OFFLINE values in the past only if they are NaN
+        # We cannot change ONLINE values in the past
+        # We can add a new date(row) in the past(ONLINE) as long as it is the same value
         logging.info('Reading old values for comparison')
-        df2 = self.read(correction)
-
-        if df2 is not None:
-            logging.info('Checking if columns unchanged in past')
+        df_old = self.read(correction)
+        if df_old is not None:
             now = datetime.now(tz=timezone.utc)
-            for column in df2.columns:
-                logging.debug(f'Checking {column}')
-                if not (df2.loc[df2.index < now, column] ==
-                        df.loc[df.index < now, column]).all():
-                    raise ValueError(f'{column} changed in past, not allowed')
+            new_dates = df.index.difference(df_old.index)
+            if not new_dates.empty:
+                for i, item in enumerate(new_dates):
+                    if item < now:
+                        for column in df_old.columns:
+                            new_value = df.loc[df.index == new_dates[i].to_pydatetime(), column].values[0]
+                            old_value = df_old.loc[df_old.index < new_dates[i].to_pydatetime(), column][-1]
+                            if new_value != old_value:
+                                raise ValueError(f'{column} changed in past, not allowed')
+            else:
+                for column in df_old.columns:
+                    if 'ONLINE' in column:
+                        if not (df.loc[df.index < now, column] == df_old.loc[df_old.index < now, column]).all():
+                            raise ValueError(f'{column} changed in past, not allowed')
+                    else:
+                        for i, item in enumerate(new_dates):
+                            old_value = df.loc[df.index < new_dates[i].to_pydatetime(), column][-1]
+                            if not np.isnan(old_value).all():
+                                raise ValueError(f'{column} only NaN values can be updated')
+
```

Comment:
I am sorry but I do not understand this logic here. To me it seems to be way to complicated. Further, I do not understand how you can change values in the past. Should not the ONLINE-part not always raise an ValueError in this case? Or is "correction" a string like "ONLINE_gain" or just "gain" and you get all columns "ONLINE", "v1", "v2" etc.?  Could you maybe illustrate with a sketch and explain in words what you would like to achieve?

Reply:
This logic is meant to allow changes to a pandas.dataframe, I can mentioned a few scenarios: the user wants to update the datetime (index) according to the actual time of the event e.g. (gain calibration time) this implies adding a new row. Another scenario, filling out non-physical values  (NaNs). 
In both cases this logic checks that 1) changes in the past are not allowed for ONLINE and only non-physical values can be modify in the past